### PR TITLE
Use inspect in assertion diff for structs if possible

### LINKED ISSUE #6930: Use inspect in assertion diff for structs
### Environment

* Elixir & Erlang versions (elixir --version): Elixir 1.6.0-dev
* Operating system: macOS Sierra

### Current behavior

On our project we've been recently dealing a lot with time and on large lists the diff is hard to read. We ended up explicitly using inspect for assertions (as shown in expected behavior section). Looks like in https://github.com/elixir-lang/elixir/pull/4430#issuecomment-202370016 `Inspect` was considered to diff in structs. Perhaps it could be reconsidered?

```
t1 = ~N[2017-10-15 09:00:00]
t2 = ~N[2017-10-15 12:00:00]
t3 = ~N[2017-10-16 22:00:00]

assert [t1, t2, t3, t1] == [t1, t1, t1, t1, t1]
```

zrzut ekranu 2017-10-15 o 10 40 40

### Expected behavior

zrzut ekranu 2017-10-15 o 10 46 50

### PR DESCRIPTION
**Pull request by wojtekmach**
Closes #6930

For this:

```elixir
  test "1" do
    assert ~N[2017-10-01 09:00:00] == ~N[2017-11-02 09:00:00.0]
  end

  test "2" do
    assert MapSet.new([1]) == MapSet.new([1, 2])
  end

  test "3" do
    assert 1..1 == 1..2
  end

  test "4" do
    assert Stream.filter([1, 2, 3], &(&1)) == Stream.take([1, 2, 3], 2)
  end
```

Before:

screen shot 2017-11-17 at 23 19 58

After:

screen shot 2017-11-17 at 23 20 26

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/diff.ex [hunk 1]
```diff
@@ -24,9 +24,13 @@ defmodule ExUnit.Diff do
 
   # Structs
   def script(%name{} = left, %name{} = right) do
-    left = Map.from_struct(left)
-    right = Map.from_struct(right)
-    script_map(left, right, inspect(name))
+    if Inspect.impl_for(left) != Inspect.Any and inspect(left) != inspect(right) do
+      script_string(inspect(left), inspect(right))
```
Comment (ericmj) [inline lib/ex_unit/lib/ex_unit/diff.ex]: We should reduce the number of times we call `inspect`.

Comment (wojtekmach) [reply lib/ex_unit/lib/ex_unit/diff.ex]: yup, I just wasn't sure how to do it elegantly. How about:

```elixir
  def script(%name{} = left, %name{} = right) do
    inspect_left = inspect(left)
    inspect_right = inspect(right)

    if Inspect.impl_for(left) != Inspect.Any and inspect_left != inspect_right do
      script_string(inspect_left), inspect_right)
    else
      left = Map.from_struct(left)
      right = Map.from_struct(right)
      script_map(left, right, inspect(name))
    end
  end
```

or

```elixir
  def script(%name{} = left, %name{} = right) do
    if Inspect.impl_for(left) != Inspect.Any
      inspect_left = inspect(left)
      inspect_right = inspect(right)

      if inspect_left != inspect_right do
        script_string(inspect_left), inspect_right)
      else
        script_struct(left, right)
      end
    else
      script_struct(left, right)
    end
  end

  defp script_struct(%name{} = left, %name{} = right) do
    left = Map.from_struct(left)
    right = Map.from_struct(right)
    script_map(left, right, inspect(name))
  end
```

Comment (ericmj) [reply lib/ex_unit/lib/ex_unit/diff.ex]: Second looks good to me.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**